<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="robot">
<xacro:include filename="inertial_macros.xacro"/>

<material name="white">
  <color rgba="1 1 1 1" />
</material>
<material name="orange">
  <color rgba="1 0.3 0.1 1"/>
</material>
<material name="blue">
  <color rgba="0.2 0.2 1 1"/>
</material>
<material name="black">
  <color rgba="0 0 0 1"/>
</material>
<material name="red">
  <color rgba="1 0 0 1"/>
</material>

<link name="base_link">
  
</link>

<!-- BASE_FOOTPRINT LINK -->
<joint name="base_footprint_joint" type="fixed">
        <parent link="base_link"/>
        <child link="base_footprint"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </joint>

    <link name="base_footprint">
    </link>

<joint name="chassis_joint" type="fixed">
  <parent link="base_link"/>
  <child link="chassis"/>
  <origin xyz="0 0 0"/>
</joint>

<link name="chassis">
  <visual>
    <origin xyz="0 0 0.075"/>
    <geometry>
      <cylinder radius="0.11" length="0.15"/>
    </geometry>
    <material name="white"/>
  </visual>
  <collision>
    <origin xyz="0 0 0.075"/>
    <geometry>
      <cylinder radius="0.11" length="0.15"/>
    </geometry>
  </collision>
  <xacro:inertial_cylinder mass="1.3" length="0.15" radius="0.11">
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </xacro:inertial_cylinder>
</link>

<joint name="left_wheel_joint" type="continuous">
  <parent link="base_link"/>
  <child link="left_wheel"/>
  <origin xyz="0 0.0875 0.0235" rpy="-${pi/2} 0 0" />
  <axis xyz="0 0 1"/>
</joint>
<link name="left_wheel">
  <visual>
    <geometry>
      <cylinder radius="0.0325" length="0.025"/>
    </geometry>
    <material name="blue"/>
  </visual>
  <collision>
    <geometry>
      <cylinder radius="0.0325" length="0.025"/>
    </geometry>
    <surface>
      <friction>
        <ode>
          <mu>1.0</mu>
          <mu2>1.0</mu2>
        </ode>
      </friction>
    </surface>
  </collision>
  <xacro:inertial_cylinder mass="0.05" length="0.025" radius="0.0325">
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </xacro:inertial_cylinder>
</link>
<gazebo reference="left_wheel">
  <mu1>1.0</mu1>
  <mu2>1.0</mu2>
</gazebo>

<joint name="right_wheel_joint" type="continuous">
  <parent link="base_link"/>
  <child link="right_wheel"/>
  <origin xyz="0 -0.0875 0.0235" rpy="${pi/2} 0 0" />
  <axis xyz="0 0 -1"/>
  <dynamics damping="0.1" friction="0.1"/>
</joint>
<link name="right_wheel">
  <visual>
    <geometry>
      <cylinder radius="0.0325" length="0.025"/>
    </geometry>
    <material name="blue"/>
  </visual>
  <collision>
    <geometry>
      <cylinder radius="0.0325" length="0.025"/>
    </geometry>
    <surface>
      <friction>
        <ode>
          <mu>1.0</mu>
          <mu2>1.0</mu2>
        </ode>
      </friction>
    </surface>
  </collision>
  <xacro:inertial_cylinder mass="0.05" length="0.025" radius="0.0325">
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </xacro:inertial_cylinder>
</link>
<gazebo reference="right_wheel">
  <mu1>1.0</mu1>
  <mu2>1.0</mu2>
</gazebo>

<joint name="caster_wheel_joint" type="fixed">
  <parent link="chassis"/>
  <child link="caster_wheel"/>
  <origin xyz="0.08 0 0.001"/>
</joint>
<link name="caster_wheel">
  <visual>
    <geometry>
      <sphere radius="0.01"/>
    </geometry>
    <material name="black"/>
  </visual>
  <collision>
    <geometry>
      <sphere radius="0.01"/>
    </geometry>
    <surface>
      <friction>
        <ode>
          <mu>0.0</mu>
          <mu2>0.0</mu2>
        </ode>
      </friction>
    </surface>
  </collision>
  <xacro:inertial_sphere mass="0.2" radius="0.01">
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </xacro:inertial_sphere>
</link>
<gazebo reference="caster_wheel">
  <mu1>0.001</mu1>
  <mu2>0.001</mu2>
</gazebo>

<transmission name="left_wheel_trans">
  <type>transmission_interface/SimpleTransmission</type>
  <joint name="left_wheel_joint">
    <hardwareInterface>PositionJointInterface</hardwareInterface>
  </joint>
  <actuator name="left_wheel_motor">
    <mechanicalReduction>1</mechanicalReduction>
  </actuator>
</transmission>
<transmission name="right_wheel_trans">
  <type>transmission_interface/SimpleTransmission</type>
  <joint name="right_wheel_joint">
    <hardwareInterface>PositionJointInterface</hardwareInterface>
  </joint>
  <actuator name="right_wheel_motor">
    <mechanicalReduction>1</mechanicalReduction>
  </actuator>
</transmission>
</robot>
